### PR TITLE
171 add dynamic info session signup for external forms

### LINKED ISSUE #171: Add dynamic info session signup for external forms
Allow signing up for info session from 3rd party forms ([facebook instant forms](https://www.facebook.com/business/help/761812391313386?id=735435806665862))

## Tasks
- Restructure website api to more clearly 
  - Rename: ~`/api/signup/*`~ → `/api/signups/*`
  - Move: ~`/api/infoSession/user`~ → `/api/signups/info` (`/index.ts`)
    - Update request on the client
  - Add: `/api/signups/info/facebook` (`facebook.ts`)

### New Structure
```
./pages/api/signups/
  ├ highschool.ts
  ├ info/
    ├ index.ts     # Root endpoint: POST /api/signups/info
    ├ facebook.ts
```

### Payload/Fields for `/api/signups/info/facebook`
```ts
export type FacebookPayloadFieldKey =
  | 'day'
  | 'time'
  | 'name'
  | 'email'
  | 'phone'
  | 'zip';

export type FacebookPayloadField = {
  name: FacebookPayloadFieldKey;
  values: string[];
};

export type InfoSessionFacebookPayload = {
  /** @example '2015-02-28T08:49:14+0000' */
  created_time: string;
  /** Lead ID */
  id: string;
  /** Ad ID */
  ad_id: string;
  /** Form ID */
  form_id: string;
  /** User Responses */
  field_data: FacebookPayloadField[];
};
```

### Resources
- [Lead ads docs](https://developers.facebook.com/docs/marketing-api/guides/lead-ads)
  - [Create form](https://developers.facebook.com/docs/marketing-api/guides/lead-ads/create)
  - [Retrieving forms](https://developers.facebook.com/docs/marketing-api/guides/lead-ads/retrieving)
  - [Facebook Webhooks](https://developers.facebook.com/docs/graph-api/webhooks/)
  - [Webhooks (retrieving)](https://developers.facebook.com/docs/marketing-api/guides/lead-ads/retrieving#webhooks)
  - [Webhooks for leads](https://developers.facebook.com/docs/marketing-api/guides/lead-ads/quickstart/webhooks-integration/)
  - [Leadgen Forms](https://developers.facebook.com/docs/graph-api/reference/page/leadgen_forms/)
  - [Subscribed apps reference](https://developers.facebook.com/docs/graph-api/reference/page/subscribed_apps)
  - [Testing tool](https://developers.facebook.com/tools/lead-ads-testing) -- TODO: Figure out how to subscribe to webhook events for lead ads
    - [Testing tool tutorial](https://leadsync.me/blog/test-facebook-lead-ads/)
  - [Instant forms](https://www.facebook.com/business/help/761812391313386?id=735435806665862)
  - [Operation Spark Facebook Forms](https://business.facebook.com/latest/instant_forms/forms?asset_id=725052907563101&business_id=1717012455274208)
  - [Operation Spark Facebook Forms (Drafts)](https://business.facebook.com/latest/instant_forms/draft_forms?asset_id=725052907563101&business_id=1717012455274208)
- [View app/user access tokens](https://developers.facebook.com/tools/accesstoken)
  - [Creating page access tokens?](https://developers.facebook.com/docs/facebook-login/guides/access-tokens#pagetokens)
  - [Create `System User Access Tokens`](https://www.facebook.com/business/help/503306463479099?id=2190812977867143) -- ***SOLUTION***
  - [Long lived page access tokens](https://developers.facebook.com/docs/facebook-login/guides/access-tokens/get-long-lived#get-a-long-lived-page-access-token)
- [Facebook Graph API docs](https://developers.facebook.com/docs/graph-api/overview) 
- [Graph API Testing tool](https://developers.facebook.com/tools/explorer/963941362163995/)
- [Operation Spark Forms App](https://developers.facebook.com/apps/963941362163995/dashboard/?business_id=1717012455274208)
- [Facebook Pages (Admin view)](https://business.facebook.com/settings/pages/725052907563101?business_id=1717012455274208)
- [Facebook Permissions](https://developers.facebook.com/docs/permissions)
- [OpSpark profile info (transparency)](https://www.facebook.com/opspark/about_profile_transparency)

### Future
- [Dynamically update the form fields](https://developers.facebook.com/docs/marketing-api/guides/lead-ads/create#lead_form) with the realtime info sessions

### PR DESCRIPTION
**Pull request by ptbarnum4**
# Dynamic info session signup API

> Reference issue https://github.com/OperationSpark/operationspark-org-website-2022/issues/171 for more details

Refactor for better structure to support signups from 3rd party sources (webhooks). This PR will focus on setting up to support incoming webhooks from facebook for `page` `leadgen` events. This API will transform the form responses and submit the forms to [**OperationSpark/service-signups**](https://github.com/OperationSpark/service-signups)

### TODO (Upon Deployment)

- [ ] Update [webhook URL](https://developers.facebook.com/apps/963941362163995/webhooks/?business_id=1717012455274208)
  ```diff
  - https://opspark.ngrok.io/api/signups/info/facebook
  + https://www.operationspark.org/api/signups/info/facebook
  ```

## Complete
- Resolves #171 
- [x] Create facebook app (Must be live in order to test)
- [x] Configure facebook lead (Instant form) -- (Must be published in order to test)
- [x] Setup API for webhook -- POST `/api/signups/info/facebook`
  - [x] Accept initial webhook subscription, validate webhook token matches, and respond with challenge code when provided the following query params:
    - `hub.mode`
    - `hub.challenge`
    - `hub.verify_token`
  - [x] For every other incoming webhook, verify the request `body` + `APP_SECRET` matches the `x-hub-signature-256` header
    ```ts
    const providedHash = req.headers['x-hub-signature-256'].split('=')[1];
    if (!providedHash) return false;

    const createdHash = crypto
      .createHmac('sha256', FB_APP_SECRET)
      .update(rawBody)
      .digest('hex');
    const isAuthReq = createdHash === providedHash;
    return isAuthReq;
    ```
  - [x] Parse the lead ids
  - [x] Fetch form data with lead ids
  - [x] Parse form data
    - [x] Find best session when provided a day and timeOfDay
    - [x] Format in the same shape of existing signups
  - [x] Update [.env.example](https://github.com/OperationSpark/operationspark-org-website-2022/blob/main/.env.example) with new vars
- [x] Add validation for possibility of webhook with invalid data
- [x] Add tests to check that the correct session is returned when "finding best session"
- [x] [Add env variables to vercel](https://vercel.com/operation-spark/opspark-website-2022/settings/environment-variables)
  - [x] `FB_PAGE_ID`
  - [x] `FB_APP_ID`
  - [x] `FB_ACCESS_TOKEN`
  - [x] `FB_APP_SECRET`
  - [x] `FB_APP_TOKEN`
  - [x] `FB_CLIENT_TOKEN`
  - [x] `FB_WEBHOOK_TOKEN` → `openssl rand -base64 32`
- [x] Add tests to validate accepting initial webhook subscription)
- [x] Add tests to verify accepting webhook (and initial subscription)
- [x] Submit form data to signup service

### CODE REVIEW

File: src/api/helpers/facebookWebhook.ts [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+export const fetchLead = async (id: string): Promise<InfoSessionFacebookPayload | null> => {
+  try {
+    const { data } = await fb.get(`/${id}`, {
+      params: {
+        access_token: FB_ACCESS_TOKEN,
+        fields: ['created_time', 'id', 'ad_id', 'form_id', 'field_data'].join(','),
+      },
+    });
+
+    return data ?? null;
```
Comment (harveysanders) [inline src/api/helpers/facebookWebhook.ts:87]:
```suggestion
    return data ?? null;
```

**Review by harveysanders — APPROVED** [review]
I left a few comments, but looks good overall. I think we should deal with this sooner than later though:
https://github.com/OperationSpark/operationspark-org-website-2022/pull/172/files/4c5550df50d4602705bba0c4b48858cfdfd6ca54#r1788229216

File: src/api/helpers/facebookWebhook.ts [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+  const payloadSha = crypto.createHmac('sha256', FB_APP_SECRET).update(rawBody).digest('hex');
+
+  const headerSha = (req.headers['x-hub-signature-256'] as string)?.split('=')[1];
+
+  if (headerSha !== payloadSha) {
+    console.error('Invalid Signature', {
+      headers: req.headers,
+      query: req.query,
+    });
+    return {
```
[the remark this answers is not in the document]
Comment (ptbarnum4) [reply src/api/helpers/facebookWebhook.ts]: updated here https://github.com/OperationSpark/operationspark-org-website-2022/pull/172/commits/b60fcc8e8cf4c9345a5f37ab0e39038c6706b87b

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -1,13 +1,13 @@
-name: e2e
+name: Cypress e2e Tests
 on: [push]
 jobs:
   cypress-run:
     name: Cypress run
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```
[the remark this answers is not in the document]
Comment (ptbarnum4) [reply .github/workflows/e2e.yml]: updated both https://github.com/OperationSpark/operationspark-org-website-2022/pull/172/commits/33ebcf83d32c4b976c48f5d149fd222e8205fd6c

File: pages/api/signups/info/facebook.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  if (body.object !== 'page') {
+    res.status(400).json({ message: 'Request not valid' });
+    return;
+  }
+  // Flatten all lead IDs
+  const leadIds = body.entry?.flatMap(
+    (entry) => entry.changes.flatMap((change) => change.value.leadgen_id) ?? [],
+  );
+  // Fetch lead form data
+  const data = await Promise.all(leadIds.map(fetchLead));
```
[the remark this answers is not in the document]
Comment (ptbarnum4) [reply pages/api/signups/info/facebook.ts]: Updated here https://github.com/OperationSpark/operationspark-org-website-2022/pull/172/commits/0b6d16b9b2173ff3f140a9adb14d36fb7f4ec04d

**Review by harveysanders — APPROVED** [review]

**Pull request merged**